runner/websocket.go

```diff
@@ -0,0 +1,131 @@
+	"google.golang.org/protobuf/proto"
+)
+
+func SetupWebsocketForAllServers(ctx context.Context, servers []*state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) {
+	for idx := range servers {
+		go func(server *state.Server) {
+			logger = logger.WithPrefixAndRememberErrors(server.Config.SectionName)
+			for {
+				if server.WebSocket.Load() == nil {
+					connect(ctx, server, globalCollectionOpts, logger)
+				}
+				time.Sleep(3 * 60 * time.Second) // Delay between reconnect attempts
+			}
+		}(servers[idx])
+	}
+}
+
+func connect(ctx context.Context, server *state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) {
+	connCtx, cancelConn := context.WithCancel(ctx)
+	proxyConfig := httpproxy.Config{
+		HTTPProxy:  server.Config.HTTPProxy,
+		HTTPSProxy: server.Config.HTTPSProxy,
+		NoProxy:    server.Config.NoProxy,
+	}
+	dialer := websocket.Dialer{
+		ReadBufferSize:  10240,
+		WriteBufferSize: 10240,
+		Proxy: func(req *http.Request) (*url.URL, error) {
+			return proxyConfig.ProxyFunc()(req.URL)
+		},
+	}
+	url, _ := url.Parse(server.Config.APIBaseURL + "/v2/snapshots/websocket")
+	if url.Scheme == "http" {
+		url.Scheme = "ws"
+	} else {
+		url.Scheme = "wss"
+	}
+	headers := make(map[string][]string)
+	headers["Pganalyze-Api-Key"] = []string{server.Config.APIKey}
+	headers["Pganalyze-System-Id"] = []string{server.Config.SystemID}
+	headers["Pganalyze-System-Type"] = []string{server.Config.SystemType}
+	headers["Pganalyze-System-Scope"] = []string{server.Config.SystemScope}
+	headers["Pganalyze-System-Id-Fallback"] = []string{server.Config.SystemIDFallback}
+	headers["Pganalyze-System-Type-Fallback"] = []string{server.Config.SystemTypeFallback}
+	headers["Pganalyze-System-Scope-Fallback"] = []string{server.Config.SystemScopeFallback}
+	headers["User-Agent"] = []string{util.CollectorNameAndVersion}
+	conn, response, err := dialer.DialContext(connCtx, url.String(), headers)
+	if err != nil {
+		cancelConn()
+		logger.PrintWarning("Error starting websocket: %s %v", err, response)
+		return
+	}
+	server.WebSocket.Store(conn)
+	server.Pause.Store(&pganalyze_collector.ServerMessage_Pause{Pause: false})
+	// Writer goroutine
+	go func() {
+		for {
+			select {
+			case <-connCtx.Done():
+				socket := server.WebSocket.Swap(nil)
+				if socket != nil {
+					err = socket.Close()
+					if err != nil {
+						logger.PrintWarning("Error closing websocket: %s", err)
+					}
+				}
+				return
+			case snapshot := <-server.SnapshotStream:
+				logger.PrintVerbose("Uploading snapshot to websocket")
+				err = server.WebSocket.Load().WriteMessage(websocket.BinaryMessage, snapshot)
+				if err != nil {
+					logger.PrintError("Error uploading snapshot: %s", err)
+					cancelConn()
+					return
+				}
+			}
+		}
+	}()
+	// Reader goroutine
+	go func() {
+		for {
+			_, compressedData, err := conn.ReadMessage()
+			if err != nil {
+				if !websocket.IsCloseError(err, 1005) { // Normal close event
```

Comment:
1005 and 1006 are the only close events I've been able to cause from the server side. The server-side websocket implementation doesn't seem to let us provide a close message.

1006 (abnormal closure) happened before the recent server-side commit that added graceful shutdown.

Reply:
Note that the event codes are from the WebSocket spec not from gorilla/websocket, so if we link to something it might be better to use https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent/code for example.

Reply:
I'm not Maciek but I _think_ what Maciek meant was changing this to below?

```suggestion
				if !websocket.IsCloseError(err, websocket.CloseNoStatusReceived) { // Normal close event
```